lib/network/internal.nix

```diff
@@ -0,0 +1,220 @@
+    */
+    _encode =
+      num:
+      if num < 0 then
+        throw "lib.network.ipv4._encode: ${toString num} cannot be encoded into an IPv4 address."
+      else if num > 4294967295 then
+        throw "lib.network.ipv4._encode: ${toString num} is too large to encode into an IPv4 address."
+      else
+        concatStringsSep "." (
+          map (x: toString (mod (num / x) 256)) (reverseList (genList (x: common.pow 2 (x * 8)) 4))
+        );
+
+    /**
+      Extracts a prefix length from a CIDR and verifies it is valid.
+
+      If an IP address is given without a CIDR, then a prefix length of 32 is returned.
+
+      # Example
+
+      ```nix
+      _verifyPrefixLength "192.168.0.1/24"
+      => 24
+      _verifyPrefixLength "192.168.0.1"
+      => 32
+      ```
+
+      # Type
+
+      ```
+      _verifyPrefixLength :: String => String
+      ```
+
+      # Arguments
+
+      - [cidr] An IPv4 CIDR.
+
+      # Throws
+
+      - If a CIDR was given with a slash but no prefix length following.
+      - If there were multiple slashes in the CIDR.
+    */
+    _verifyPrefixLength =
+      cidr:
+      let
+        splitCidr = lib.splitString "/" cidr;
+      in
+      if (builtins.length splitCidr) == 1 then
+        "32"
+      else if (builtins.length splitCidr) > 2 then
+        throw "lib.network.ipv4: Could not verify prefix length for CIDR ${cidr}."
+      else
+        let
+          afterSlash = elemAt splitCidr 1;
+        in
+        if afterSlash == "" then
+          throw "lib.network.ipv4: CIDR ${cidr} has no prefix length."
+        else if toInt afterSlash > 32 || toInt afterSlash < 0 then
+          throw "lib.network.ipv4: CIDR ${cidr} has an out of bounds prefix length, ${afterSlash}."
+        else
+          afterSlash;
+
+    _verifyAddress =
+      cidr:
+      let
+        splitCidr = lib.splitString "/" cidr;
+        address = elemAt splitCidr 0;
+        splitAddress = lib.splitString "." address;
+        intOctets = map toInt splitAddress;
+      in
+      if (builtins.length splitAddress) != 4 then
+        throw "lib.network.ipv4: CIDR ${cidr} is not of the correct form."
+      else if any (x: x == "") splitAddress then
+        throw "lib.network.ipv4: CIDR ${cidr} has an empty octet."
+      else if !(all (x: x >= 0 && x <= 255)) intOctets then
+        throw "lib.network.ipv4: CIDR ${cidr} has an out of bounds octet."
+      else
+        address;
+
+    /**
+      Given an IP address and prefix length, creates an attribute set of network parameters.
+
+      # Example
+
+      ```nix
+      _parse "192.168.0.1/24"
+      => {
+        address = "192.168.0.1";
+        cidr = "192.168.0.1/24";
+        prefixLength = "24";
+        subnetMask = "255.255.255.0";
+      }
+      ```
+
+      # Type
+
+      ```
+      _parse :: String -> IPv4Address
+      ```
+
+      # Arguments
+
+      - [address] An IPv4 address.
+      - [prefixLength] A prefix length.
+    */
+    _parse = address: prefixLength: {
```

Comment:
This function doesn't actually do any parsing at all. Overall I think these internal functions could be refactored to this:
- `_parseIp :: String -> String -> Int`
  Example:
  ```nix
  _parseIp "lib.network.ipv4.someFunction" "12.34.56.78"
  -> 203569230
  
  _parseIp "lib.network.ipv4.someFunction "12.34.56"
  -> error: lib.network.ipv4.someFunction: IPv4 address "12.34.56" is not of the correct form
  ```
- `_parsePrefix :: String -> String -> Int`
  Example:
  ```nix
  _parsePrefix "lib.network.ipv4.someFunction" "16"
  -> 16
  
  _parsePrefix "lib.network.ipv4.someFunction" "160"
  -> error: lib.network.ipv4.someFunction: IPv4 prefix "160" is too large.
  ```
- `_parseCIDR :: String -> String -> { ip :: Int, prefix :: Int }`
  Splits and calls `_parseIp` and `_parsePrefix` underneath

At this point, the public `fromCidrString` function can call `_parseCIDR "lib.network.ipv4.fromCidrString"` to get back two integers that are _guaranteed_ to be valid. If they're not you'll get failure that also indicates that function call failed (you can add extra context to the error messages as desired).

All the internal functions for _generating_ data now don't need to do any validation themselves.

Reply:
I'm having a hard time understand your suggestion. What is the first argument for these parse functions supposed to do?

For example, in `_parseIp "lib.network.ipv4.someFunction" "12.34.56.78"`, I don't understand what the intent of passing `"lib.network.ipv4.someFunction"` and how it would be used in the function.